server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferMetadata.java

```diff
@@ -122,6 +142,25 @@ private void write(DataOutput out) throws IOException {
         out.writeLong(generation);
         out.writeLong(minTranslogGeneration);
         out.writeLong(timeStamp);
-        out.writeMapOfStrings(generationToPrimaryTermMapper.get());
+        if (generationToPrimaryTermMapper.get() != null) {
+            out.writeMapOfStrings(generationToPrimaryTermMapper.get());
+        } else {
+            out.writeMapOfStrings(new HashMap<>());
+        }
+    }
+
+    private static class MetadataFilenameComparator implements Comparator<String> {
+        @Override
+        public int compare(String metadaFilename1, String metadaFilename2) {
+            // Format of metadata filename is <Primary Term>__<Generation>__<Timestamp>
+            String[] filenameTokens1 = metadaFilename1.split(METADATA_SEPARATOR);
+            String[] filenameTokens2 = metadaFilename2.split(METADATA_SEPARATOR);
+            for (int i = 0; i < filenameTokens1.length; i++) {
+                if (filenameTokens1[i].equals(filenameTokens2[i]) == false) {
+                    return (int) (Long.parseLong(filenameTokens1[i]) - Long.parseLong(filenameTokens2[i]));
+                }
+            }
+            return 0;
+        }
```

Comment:
can we add a WARN log here - we should not come across such situation where 2 metadata file comparison yield 0. Also, could we add log when we have to resolve to timestamp for comparing? I dont think using timestamp is fair as clocks are not synchronised across the nodes in the distributed system setup.

Reply:
Totally agree with @ashking94's concerns here. However, putting WARN statements in a comparator could lead to craziness in the log. We don't really have control over how many times and against which pairs of files the sort algorithm will invoke the comparison. Is there somewhere else we can add these statements?

Reply:
I was also not inclined to add the log in the `compare` method as it may not provide good debug insights. But as I think more, both the cases: comparing timestamp as well as returning 0 should happen in exceptional cases (we think it should not happen at all). Also, it is not feasible to add this logic in upload flow as it means reading the last uploaded file each time. IMO, we should add these logs in this method. Thoughts?

Reply:
`2 metadata file comparison yield 0.` : This will never happen , as there can't be two files with same name in a remote directory . 

Timestamp comparison is a tricky thing here - We can check this in upload flow as it has `FileTransferTracker` . However , we don't add metadata files in that . Alternate is  we just throw `RuntimeException`  from `readMetadata()` ? This is better than logs (we can miss it ) and silent failures

Reply:
The expectation is that two files with the same primary term and generation should never exist, right? If so, then can we just not include the timestamp in the filename? Then it would fail at upload time, when the file was generated that violated the expected invariant, which seems better than failing hard here (because I think the system would end up stuck because it would keep hitting this same error until something changed).

Reply:
@sachinpkale in case of a primary-primary relocation, the <primary term>_<generation> can be same or will be different for the xlog upload? If yes, then failing upload might as well be not correct.

Reply:
>The expectation is that two files with the same primary term and generation should never exist, right
This should be extremely rare but should still technically possible for isolated writers during primary-primary relocation like what @sachinpkale mentioned. The timestamp serves as a discriminator in all those cases, based on a last writer wins policy.